tracing-appender/src/rolling.rs

```diff
@@ -253,21 +227,32 @@ impl RollingFileAppender {
 
 impl io::Write for RollingFileAppender {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let should_rollover = |now: OffsetDateTime| -> bool {
+            if let Some(current_time) = self.state.should_rollover_by_date(now) {
+                let did_cas = self.state.advance_date(now, current_time);
+                debug_assert!(did_cas, "if we have &mut access to the appender, no other thread can have advanced the timestamp...");
+                return true;
+            }
+            if self.state.should_rollover_by_size() {
+                self.state.advance_same_date_filename_index();
+                return true;
+            }
+            false
+        };
+
         let now = self.now();
+        let should_rollover = should_rollover(now);
```

Comment:
style: It looks like we don’t need to use a closure at this place. This is making the code less readable in my opinion. We also don’t actually need to short-circuit the code using `return` because `if`-blocks are also expressions:

```suggestion
        let now = self.now();

        let should_rollover = if let Some(current_time) = self.state.should_rollover_by_date(now) {
            let did_cas = self.state.advance_date(now, current_time);
            debug_assert!(did_cas, "if we have &mut access to the appender, no other thread can have advanced the timestamp...");
            true
        } else if self.state.should_rollover_by_size() {
            self.state.advance_same_date_filename_index();
            true
        } else {
            false
        };
```

Reply:
True, this is a leftover from a previous version where I needed the `writer` inside the `should_rollover` logic, and I added the closure to handle the writer lock.